aiohttp/client.py

```diff
@@ -218,6 +226,22 @@ def _request(self, method, url, *,
         handle = tm.start()
 
         url = URL(url)
+
+        if self._trace_config:
+            if trace_context is None:
+                trace_context = self._trace_config.trace_context()
+            trace = Trace(self._trace_config, self, trace_context)
+        else:
+            trace = None
+
+        if trace:
+            yield from trace.send(
+                'on_request_start',
```

Comment:
Can we avoid strings here and have `trace.on_request_start` and friends which actually does `trace.send`, but with proper event name (`'on_request_start'` in this case)?

Reply:
The `Trace` acts as a proxy to call the signal with the proper arguments, aside of the arguments provided by the same signal such as `metod`, `url` and `headers` the `session` and the `trace_context` are sent implicitly by the `Trace.send` method. 

I came up with another implementation, see the following snippet, but I didn't like it::

```python
class Trace:
    ....

    async def on_request_start(self, *args, **kwargs):
        return await self._trace_config.on_request_start.send(
            self._session,
            self._trace_context,
            *args,
            **kwargs
        )
```

IMHO this implementation raises another issue, the class exposes a method `on_request_start` as it would be a typical signal, meanwhile is just a method that wrappers the real signal.

I know that `Trace` is not a first level class and we can do whatever we want but I have preferred to leave something consistency than having something almost equal but not consistency.

Anyway, I'm open to suggestions.

Reply:
I didn't get the issue with the snippet, could you elaborate a bit more, please?

Reply:
The usual way that the signals are exposed in aiohttp follows the following pattern::

```python
class Foo:
    def __init__(self):
        self._on_some_event = Signal(self)

    @property
    def on_some_event(self):
        return self._on_some_event

f = Foo()
f.on_some_event.send(arg1, arg2)
```

The interface of the signal is fully exposed, its responsibility of the caller to call it using the `send` method and the arguments to trigger all listeners for a specific event.

The previous snippet shows how the signal is not fully exposed and the `send` method is hidden to allow the `Trace` object add the two first params. 

IMHO the snippet implementation is kinda inconsistent if you compare it with the traditional way of exposing signals. 

To avoid that misinterpretation I preferred to create an ad-hoc method called `send` for the `Trace` object.

Reply:
What's about `Trace.send_request_start()` method instead of `on_request_start`?

Reply:
Good idea, Andrew!